cove_bods/process.py

```diff
@@ -0,0 +1,395 @@
+        if not os.path.exists(self.data_filename):
+            save_data = {
+                "schema_version_used": process_data['schema'].schema_version
+            }
+            with open(self.data_filename, "w") as fp:
+                json.dump(save_data, fp, indent=4)
+        # return
+        return process_data
+
+    def get_context(self):
+        context = {}
+        # data
+        if os.path.exists(self.data_filename):
+            with open(self.data_filename) as fp:
+                context.update(json.load(fp))
+        # done!
+        return context
+
+
+CONVERT_JSON_INTO_SPREADSHEETS_DIR_NAME = "flatten"
+
+
+class ConvertJSONIntoSpreadsheets(ProcessDataTask):
+    """Convert primary format (JSON) to spreadsheets"""
+
+    def __init__(
+        self, supplied_data: SuppliedData, supplied_data_files: List[SuppliedDataFile]
+    ):
+        super().__init__(supplied_data, supplied_data_files)
+        self.output_dir = os.path.join(
+            self.supplied_data.data_dir(),
+            CONVERT_JSON_INTO_SPREADSHEETS_DIR_NAME,
+            "data",
+        )
+        self.xlsx_filename = os.path.join(
+            self.supplied_data.data_dir(),
+            CONVERT_JSON_INTO_SPREADSHEETS_DIR_NAME,
+            "data.xlsx",
+        )
+
+    def is_processing_applicable(self) -> bool:
+        return True
+
+    def is_processing_needed(self) -> bool:
+        return not os.path.exists(self.xlsx_filename)
+
+    def process(self, process_data: dict) -> dict:
+
+        # don't run if already done
+        if os.path.exists(self.xlsx_filename):
+            return process_data
+
+        os.makedirs(self.output_dir, exist_ok=True)
+
+        flatten_kwargs = {
+            "output_name": self.output_dir,
+            "root_list_path": "there-is-no-root-list-path",
+            "root_id": "statementID",
+            "id_name": "statementID",
+            "root_is_list": True,
+            "schema": process_data['schema'].pkg_schema_url,
+        }
+
+        try:
+            flattentool.flatten(process_data["json_data_filename"], **flatten_kwargs)
+        except Exception as err:
+            capture_exception(err)
+            # TODO log and show to user. https://github.com/Open-Telecoms-Data/cove-ofds/issues/24
+
+        return process_data
+
+    def get_context(self):
+        context = {}
+        # XLSX
+        if os.path.exists(self.xlsx_filename):
+            context["can_download_xlsx"] = True
+            context["download_xlsx_url"] = os.path.join(
+                self.supplied_data.data_url(),
+                CONVERT_JSON_INTO_SPREADSHEETS_DIR_NAME,
+                "data.xlsx",
+            )
+            context["download_xlsx_size"] = os.stat(self.xlsx_filename).st_size
+        else:
+            context["can_download_xlsx"] = False
+        # done!
+        return context
+
+
+class PythonValidateTask(TaskWithState):
+
+    state_filename: str = "python_validate.py"
```

Comment:
I'm confused as to why this has a `.py` extension. Isn't it going to have json written to it?

Reply:
Whoops